File: src/main/kotlin/com/github/dungphan/unityindex/tools/unity/GetAssemblyMapTool.kt
package com.github.dungphan.unityindex.tools.unity

import com.github.dungphan.unityindex.constants.ToolNames
import com.github.dungphan.unityindex.server.models.ToolCallResult
import com.github.dungphan.unityindex.tools.AbstractMcpTool
import com.github.dungphan.unityindex.tools.models.AssemblyDefinition
import com.github.dungphan.unityindex.tools.models.AssemblyMapResult
import com.github.dungphan.unityindex.tools.schema.SchemaBuilder
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VfsUtilCore
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileVisitor
import com.intellij.openapi.vfs.LocalFileSystem
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class GetAssemblyMapTool : AbstractMcpTool() {

    override val requiresPsiSync: Boolean = false

    override val name = ToolNames.GET_ASSEMBLY_MAP

    override val description = """
        Get the Unity project's assembly definition (.asmdef) structure and dependency graph. Returns all assembly definitions with their names, references, platform targets, and constraints.

        Use this to understand how the codebase is partitioned into compilation units, what depends on what, and where to place new code. Editor-only assemblies are flagged.

        Parameters: project_path (optional, only needed with multiple projects open).

        Example: {}
    """.trimIndent()

    override val inputSchema: JsonObject = SchemaBuilder.tool()
        .projectPath()
        .build()

    override suspend fun doExecute(project: Project, arguments: JsonObject): ToolCallResult {
        val basePath = project.basePath
            ?: return createErrorResult("Cannot determine project base path")

        val projectDir = LocalFileSystem.getInstance().findFileByPath(basePath)
            ?: return createErrorResult("Project directory not found: $basePath")

        val asmdefFiles = mutableListOf<VirtualFile>()
        VfsUtilCore.visitChildrenRecursively(projectDir, object : VirtualFileVisitor<Unit>() {
            override fun visitFile(file: VirtualFile): Boolean {
                if (file.extension == "asmdef") {
                    asmdefFiles.add(file)
                }
                return true
            }
        })

        if (asmdefFiles.isEmpty()) {
            return createJsonResult(AssemblyMapResult(
                assemblies = emptyList(),
                totalCount = 0,
                projectPath = basePath
            ))
        }

        val assemblies = asmdefFiles.mapNotNull { file ->
            parseAsmdef(file, basePath)
        }.sortedBy { it.name }

        return createJsonResult(AssemblyMapResult(
            assemblies = assemblies,
            totalCount = assemblies.size,
            projectPath = basePath
        ))
    }

    private fun parseAsmdef(file: VirtualFile, basePath: String): AssemblyDefinition? {
        return try {
            val content = String(file.contentsToByteArray(), Charsets.UTF_8)
            val jsonObj = json.parseToJsonElement(content).jsonObject

            val name = jsonObj["name"]?.jsonPrimitive?.content ?: return null
            val relativePath = file.path.removePrefix(basePath).removePrefix("/")

            val references = jsonObj["references"]?.jsonArray
                ?.map { it.jsonPrimitive.content } ?: emptyList()

            val includePlatforms = jsonObj["includePlatforms"]?.jsonArray
                ?.map { it.jsonPrimitive.content } ?: emptyList()

            val excludePlatforms = jsonObj["excludePlatforms"]?.jsonArray
                ?.map { it.jsonPrimitive.content } ?: emptyList()

            val defineConstraints = jsonObj["defineConstraints"]?.jsonArray
                ?.map { it.jsonPrimitive.content } ?: emptyList()

            val isEditorOnly = includePlatforms.singleOrNull() == "Editor" ||
                relativePath.contains("/Editor/", ignoreCase = true)

            AssemblyDefinition(
                name = name,
                file = relativePath,
                rootNamespace = jsonObj["rootNamespace"]?.jsonPrimitive?.content?.takeIf { it.isNotBlank() },
                references = references,
                includePlatforms = includePlatforms,
                excludePlatforms = excludePlatforms,
                allowUnsafeCode = jsonObj["allowUnsafeCode"]?.jsonPrimitive?.booleanOrNull ?: false,
                autoReferenced = jsonObj["autoReferenced"]?.jsonPrimitive?.booleanOrNull ?: true,
                noEngineReferences = jsonObj["noEngineReferences"]?.jsonPrimitive?.booleanOrNull ?: false,
                defineConstraints = defineConstraints,
                isEditorOnly = isEditorOnly
            )
        } catch (e: Exception) {
            null
        }
    }
}

File: vscode-extension/src/tools/unity/getAssemblyMapTool.ts
import * as fs from "fs";
import * as path from "path";
import { AbstractMcpTool } from "../abstractTool";
import { TOOL_NAMES } from "../../constants";
import { ToolCallResult } from "../../models/jsonRpc";
import { ProjectContext, toRelativePath } from "../../server/projectResolver";
import { Args } from "../../utils/args";
import { SchemaBuilder } from "../../utils/schema";

interface AssemblyDefinition {
  name: string;
  file: string;
  rootNamespace?: string;
  references: string[];
  includePlatforms: string[];
  excludePlatforms: string[];
  allowUnsafeCode: boolean;
  autoReferenced: boolean;
  noEngineReferences: boolean;
  defineConstraints: string[];
  isEditorOnly: boolean;
}

interface AssemblyMapResult {
  assemblies: AssemblyDefinition[];
  totalCount: number;
  projectPath: string;
}

const SKIP_DIRS = new Set(["Library", "Temp", "Logs", "obj", "bin", "node_modules", ".git"]);

export class GetAssemblyMapTool extends AbstractMcpTool {
  protected readonly requiresLsp = false;

  readonly name = TOOL_NAMES.GET_ASSEMBLY_MAP;
  readonly description =
    "Get the Unity project's assembly definition (.asmdef) structure and dependency graph. Returns all .asmdef files with their references, platforms, and constraints. " +
    "Use this to understand how the codebase is partitioned into compilation units.";
  readonly inputSchema = SchemaBuilder.tool().projectPath().build();

  protected async doExecute(
    project: ProjectContext,
    _args: Args,
  ): Promise<ToolCallResult> {
    const asmdefs: AssemblyDefinition[] = [];
    walk(project.rootPath, (entry, isDir) => {
      if (isDir) return !SKIP_DIRS.has(path.basename(entry));
      if (!entry.endsWith(".asmdef")) return true;
      const parsed = parseAsmdef(entry, project);
      if (parsed) asmdefs.push(parsed);
      return true;
    });

    asmdefs.sort((a, b) => a.name.localeCompare(b.name));
    const result: AssemblyMapResult = {
      assemblies: asmdefs,
      totalCount: asmdefs.length,
      projectPath: project.rootPath,
    };
    return this.json(result);
  }
}

function parseAsmdef(
  absPath: string,
  project: ProjectContext,
): AssemblyDefinition | null {
  try {
    const content = fs.readFileSync(absPath, "utf-8");
    const data = JSON.parse(content) as Record<string, unknown>;
    const name = data.name as string | undefined;
    if (!name) return null;

    const relativeFile = toRelativePath(project, absPath);
    const includePlatforms = (data.includePlatforms as string[] | undefined) ?? [];

    return {
      name,
      file: relativeFile,
      rootNamespace: typeof data.rootNamespace === "string" && data.rootNamespace.length > 0
        ? (data.rootNamespace as string)
        : undefined,
      references: (data.references as string[] | undefined) ?? [],
      includePlatforms,
      excludePlatforms: (data.excludePlatforms as string[] | undefined) ?? [],
      allowUnsafeCode: Boolean(data.allowUnsafeCode),
      autoReferenced: data.autoReferenced === undefined ? true : Boolean(data.autoReferenced),
      noEngineReferences: Boolean(data.noEngineReferences),
      defineConstraints: (data.defineConstraints as string[] | undefined) ?? [],
      isEditorOnly:
        (includePlatforms.length === 1 && includePlatforms[0] === "Editor") ||
        relativeFile.toLowerCase().includes("/editor/"),
    };
  } catch {
    return null;
  }
}

function walk(
  root: string,
  visit: (path: string, isDir: boolean) => boolean,
): void {
  let entries: string[];
  try {
    entries = fs.readdirSync(root);
  } catch {
    return;
  }
  for (const name of entries) {
    const full = path.join(root, name);
    let stat: fs.Stats;
    try {
      stat = fs.statSync(full);
    } catch {
      continue;
    }
    if (stat.isDirectory()) {
      if (visit(full, true)) walk(full, visit);
    } else {
      visit(full, false);
    }
  }
}
